airbyte-webapp/src/hooks/useAnalytics.tsx

```diff
@@ -1,25 +1,64 @@
-import React, { useContext, useMemo } from "react";
+import React, { useContext, useMemo, useState } from "react";
 import { AnalyticsService } from "core/analytics/AnalyticsService";
 
-const analyticsServiceContext = React.createContext<AnalyticsService | null>(
+export type AnalyticsServiceProviderValue = {
+  setContext: (ctx: Record<string, unknown>) => void;
+  addContextProps: (props: Record<string, unknown>) => void;
+  removeContextProps: (props: string[]) => void;
+  service: AnalyticsService;
+};
+
+const analyticsServiceContext = React.createContext<AnalyticsServiceProviderValue | null>(
   null
 );
 
 function AnalyticsServiceProvider({
   children,
-  userId,
   version,
+  context,
 }: {
   children: React.ReactNode;
   version?: string;
-  userId?: string;
+  context?: Record<string, unknown>;
 }) {
+  const [ctx, setCtx] = useState<Record<string, unknown>>(context || {});
+
   const analyticsService: AnalyticsService = useMemo(
-    () => new AnalyticsService(userId, version),
-    [version, userId]
+    () => new AnalyticsService(ctx, version),
+    [version, ctx]
   );
+
+  const handleAddContextProps = (props: Record<string, unknown>) => {
+    return {
+      ...ctx,
+      ...props,
+    };
+  };
+
+  const handleRemoveContextProps = (props: string[]) => {
+    const newCtx = Object.keys(ctx).reduce((result, key) => {
+      if (props.includes(key)) {
+        return result;
+      }
+
+      return {
+        [key]: ctx[key],
+        ...result,
```

Comment:
As you are building from new object, there is no need to spread here and just `result[key]` could be used.